lib/string_decoder.js

```diff
@@ -2,224 +2,203 @@
-  if (this.detectIncompleteChar(buffer))
-    charLength = this.charLength;
-  charReceived = this.charReceived;
-
-  var end = buflen;
-  if (charLength) {
-    // buffer the incomplete character bytes we got
-    buffer.copy(charBuffer, 0, buflen - charReceived, end);
-    end -= charReceived;
+    i = this.lastNeed;
+    this.lastNeed = 0;
+  } else {
+    i = 0;
   }
+  if (i < buf.length)
+    return (r ? r + this.text(buf, i) : this.text(buf, i));
+  return r || '';
+};
 
-  this.charLength = charLength;
-  charStr += buffer.toString(encoding, 0, end);
-
-  end = charStr.length - 1;
-  charCode = charStr.charCodeAt(end);
-  // CESU-8: lead surrogate (D800-DBFF) is also the incomplete character
-  if (charCode >= 0xD800 && charCode <= 0xDBFF) {
-    charLength += surrogateSize;
-    charReceived += surrogateSize;
-    charBuffer.copy(charBuffer, surrogateSize, 0, surrogateSize);
-    buffer.copy(charBuffer, 0, 0, surrogateSize);
+StringDecoder.prototype.end = utf8End;
 
-    this.charLength = charLength;
-    this.charReceived = charReceived;
+// Returns only complete characters in a Buffer
+StringDecoder.prototype.text = utf8Text;
 
-    return charStr.substring(0, end);
+// Attempts to complete a partial character using bytes from a Buffer
+StringDecoder.prototype.fillLast = function(buf) {
+  if (this.lastNeed <= buf.length) {
+    buf.copy(this.lastChar, this.lastTotal - this.lastNeed, 0, this.lastNeed);
+    return this.lastChar.toString(this.encoding, 0, this.lastTotal);
   }
-
-  // or just emit the charStr
-  return charStr;
+  buf.copy(this.lastChar, this.lastTotal - this.lastNeed, 0, buf.length);
+  this.lastNeed -= buf.length;
 };
 
-// detectIncompleteChar determines if there is an incomplete UTF-8 character at
-// the end of the given buffer. If so, it sets this.charLength to the byte
-// length that character, and sets this.charReceived to the number of bytes
-// that are available for this character.
-StringDecoder.prototype.detectIncompleteChar = function(buffer) {
-  var buflen = buffer.length;
-  // determine how many bytes we have to check at the end of this buffer
-  var i = (buflen >= 3) ? 3 : buflen;
-  var newlen = false;
-
-  // Figure out if one of the last i bytes of our buffer announces an
-  // incomplete char.
-  for (; i > 0; i--) {
-    var c = buffer[buflen - i];
-
-    // See http://en.wikipedia.org/wiki/UTF-8#Description
-
-    // 110XXXXX
-    if (i === 1 && c >> 5 === 0x06) {
-      this.charLength = 2;
-      newlen = true;
-      break;
+// The UTF-8 decoder optimistically converts the Buffer to UTF-8 first, then
+// looks for any invalid character bytes at the end of the resulting string.
+// It then checks how many bytes are needed based on the upper bits found in the
+// byte of the last invalid character (from the end of the string).
+function utf8Text(buf, i) {
+  const r = buf.toString('utf8', i);
+  var nb = 0;
+  while (r.charCodeAt(r.length - 1 - nb) === 65533)
```

Comment:
nit: I’d use the hex constant here… that shouldn’t make much of a difference and I’d say it’s a more easily recognized value (I’m well aware that that’s very subjective :smile:).